pkg/controller/autoops/controller.go

```diff
@@ -0,0 +1,252 @@
+// you may not use this file except in compliance with the Elastic License 2.0.
+
+package autoops
+
+import (
+	"context"
+	"fmt"
+	"sync/atomic"
+	"time"
+
+	"go.elastic.co/apm/v2"
+	corev1 "k8s.io/api/core/v1"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/client-go/tools/record"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/controller"
+	"sigs.k8s.io/controller-runtime/pkg/handler"
+	"sigs.k8s.io/controller-runtime/pkg/manager"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+	"sigs.k8s.io/controller-runtime/pkg/source"
+
+	autoopsv1alpha1 "github.com/elastic/cloud-on-k8s/v3/pkg/apis/autoops/v1alpha1"
+	esv1 "github.com/elastic/cloud-on-k8s/v3/pkg/apis/elasticsearch/v1"
+	"github.com/elastic/cloud-on-k8s/v3/pkg/controller/common"
+	commonesclient "github.com/elastic/cloud-on-k8s/v3/pkg/controller/common/esclient"
+	"github.com/elastic/cloud-on-k8s/v3/pkg/controller/common/events"
+	"github.com/elastic/cloud-on-k8s/v3/pkg/controller/common/license"
+	"github.com/elastic/cloud-on-k8s/v3/pkg/controller/common/operator"
+	"github.com/elastic/cloud-on-k8s/v3/pkg/controller/common/tracing"
+	"github.com/elastic/cloud-on-k8s/v3/pkg/controller/common/watches"
+	"github.com/elastic/cloud-on-k8s/v3/pkg/utils/k8s"
+	ulog "github.com/elastic/cloud-on-k8s/v3/pkg/utils/log"
+)
+
+const (
+	controllerName = "autoops-controller"
+)
+
+// defaultRequeue is the default requeue interval for this controller.
+var defaultRequeue = 5 * time.Second
+
+// Add creates a new AutoOpsAgentPolicy controller and adds it to the manager with default RBAC. The manager will set fields on the controller
+// and start it when the manager is started.
+func Add(mgr manager.Manager, params operator.Parameters) error {
+	r := newReconciler(mgr, params)
+	c, err := common.NewController(mgr, controllerName, r, params)
+	if err != nil {
+		return err
+	}
+	return addWatches(mgr, c, r)
+}
+
+func newReconciler(mgr manager.Manager, params operator.Parameters) *ReconcileAutoOpsAgentPolicy {
+	k8sClient := mgr.GetClient()
+	return &ReconcileAutoOpsAgentPolicy{
+		Client:           k8sClient,
+		recorder:         mgr.GetEventRecorderFor(controllerName),
+		licenseChecker:   license.NewLicenseChecker(k8sClient, params.OperatorNamespace),
+		params:           params,
+		dynamicWatches:   watches.NewDynamicWatches(),
+		esClientProvider: commonesclient.NewClient,
+	}
+}
+
+func addWatches(mgr manager.Manager, c controller.Controller, r *ReconcileAutoOpsAgentPolicy) error {
+	// watch for changes to AutoOpsAgentPolicy
+	if err := c.Watch(source.Kind(mgr.GetCache(), &autoopsv1alpha1.AutoOpsAgentPolicy{}, &handler.TypedEnqueueRequestForObject[*autoopsv1alpha1.AutoOpsAgentPolicy]{})); err != nil {
+		return err
+	}
+
+	// watch dynamically referenced secrets
+	return c.Watch(source.Kind(mgr.GetCache(), &corev1.Secret{}, r.dynamicWatches.Secrets))
+}
+
+var _ reconcile.Reconciler = &ReconcileAutoOpsAgentPolicy{}
```

Comment:
no need to allocate a value you care only about the type 🙂 
```suggestion
var _ reconcile.Reconciler = (*ReconcileAutoOpsAgentPolicy)(nil)
```